pipelines/controller/src/main/java/org/openmrs/analytics/PipelineManager.java

```diff
@@ -304,6 +306,82 @@ synchronized void runIncrementalPipeline()
     }
   }
 
+  /**
+   * This method checks upon Controller start checks on Thrift sever to create resource tables if
+   * they don't exist. There is a @PostConstruct method present in this class which is initDwhStatus
+   * and the reason below code has not been added because dataProperties.getThriftserverHiveConfig()
+   * turns out to be null when used by
+   * DatabaseConfiguration.createConfigFromFile(dataProperties.getThriftserverHiveConfig()).
+   */
+  @Override
+  public void onApplicationEvent(ApplicationReadyEvent applicationReadyEvent) {
+    if (!dataProperties.isCreateHiveResourceTables()) {
+      return;
+    }
+
+    DatabaseConfiguration dbConfig;
+    try {
+      dbConfig =
+          DatabaseConfiguration.createConfigFromFile(dataProperties.getThriftserverHiveConfig());
+    } catch (IOException e) {
+      logger.error("Exception while reading thrift hive config.");
+      throw new RuntimeException(e);
+    }
+    HiveTableManager hiveTableManager =
+        new HiveTableManager(
+            dbConfig.makeJdbsUrlFromConfig(),
+            dbConfig.getDatabaseUser(),
+            dbConfig.getDatabasePassword());
+
+    String rootPrefix = dataProperties.getDwhRootPrefix();
+    Preconditions.checkState(rootPrefix != null && !rootPrefix.isEmpty());
+
+    String prefix = getPrefix(rootPrefix);
+    String baseDir = getBaseDir(rootPrefix);
+    List<ResourceId> paths;
+    try {
+      paths =
+          getAllChildDirectories(baseDir).stream()
+              .filter(dir -> dir.getFilename().startsWith(prefix))
+              .collect(Collectors.toList());
+
+      Preconditions.checkState(paths != null, "Make sure DWH prefix is a valid path!");
+
+      int snapshotCount = 0;
+      for (ResourceId path : paths) {
+        if (!path.getFilename().startsWith(prefix + DataProperties.TIMESTAMP_PREFIX)) {
+          logger.warn(
+              "DWH directory {} does not start with {}",
+              paths,
+              prefix + DataProperties.TIMESTAMP_PREFIX);
+        }
+        snapshotCount++;
+        Set<ResourceId> childPaths = getAllChildDirectories(baseDir + "/" + path.getFilename());
+        for (ResourceId resourceId : childPaths) {
+          String resource = resourceId.getFilename();
+          String[] tokens = path.getFilename().split(prefix + DataProperties.TIMESTAMP_PREFIX);
+          if (tokens.length > 1) {
+            String timestamp = tokens[1];
+            String thriftServerParquetPath =
+                baseDir + "/" + path.getFilename() + "/" + resourceId.getFilename();
+            logger.debug("thriftServerParquetPath: ", thriftServerParquetPath);
+            hiveTableManager.createResourceTable(resource, timestamp, thriftServerParquetPath);
+            // Create Canonical table for the latest snapshot.
+            if (snapshotCount == paths.size()) {
```

Comment:
I guess the `paths` which is the list of child directories is not guaranteed to be sorted by the filename (timestamp in the filename), so this logic of relying on the last path might not be robust.

Reply:
done